connectors/sources/sharepoint_online.py

```diff
@@ -663,125 +719,268 @@ async def validate_config(self):
+            # Therefore we mix in site_list id to it to make sure they are
+            # globally unique.
+            # Also we need to remember original ID because when a document
+            # is yielded, its "id" field is overwritten with content of "_id" field
+            list_item_natural_id = list_item["id"]
+            list_item["_id"] = f"{site_list_id}-{list_item['id']}"
+            list_item["object_type"] = "list_item"
+            list_item["_tempfile_suffix"] = os.path.splitext(
+                list_item.get("FileName", "")
+            )[-1]
+
+            content_type = list_item["contentType"]["name"]
+
+            if content_type in [
+                "Web Template Extensions",
+                "Client Side Component Manifests",
+            ]:  # TODO: make it more flexible. For now I ignore them cause they 404 all the time
+                continue
+
+            if "Attachments" in list_item["fields"]:
+                async for list_item_attachment in self.client.site_list_item_attachments(
+                    site_web_url, site_list_name, list_item_natural_id
+                ):
+                    list_item_attachment["_id"] = list_item_attachment["odata.id"]
+                    list_item_attachment["object_type"] = "list_item_attachment"
+                    list_item_attachment["_timestamp"] = list_item[
+                        "lastModifiedDateTime"
+                    ]
+                    attachment_download_func = partial(
+                        self.get_attachment_content, list_item_attachment
+                    )
+                    yield list_item_attachment, attachment_download_func
+
+            yield list_item, None
+
+    async def site_lists(self, site_id):
+        async for site_list in self.client.site_lists(site_id):
+            site_list["_id"] = site_list["id"]
+            site_list["object_type"] = "site_list"
+
+            yield site_list
+
+    async def site_pages(self, url):
+        async for site_page in self.client.site_pages(url):
+            site_page["_id"] = site_page[
+                "odata.id"
+            ]  # Apparantly site_page["GUID"] is not globally unique
+            site_page["object_type"] = "site_page"
+
+            for html_field in ["LayoutWebpartsContent", "CanvasContent1"]:
+                if html_field in site_page:
+                    site_page[html_field] = html_to_text(site_page[html_field])
+
+            yield site_page
+
+    def initiate_sync_cursor(self):
+        if not self._sync_cursor:
+            self._sync_cursor = {CURSOR_SITE_DRIVE_KEY: {}}
+
+        return self._sync_cursor
+
+    def update_drive_delta_link(self, drive_id, link):
+        self._sync_cursor[CURSOR_SITE_DRIVE_KEY][drive_id] = link
+
+    def get_drive_delta_link(self, drive_id):
+        return self._sync_cursor.get(CURSOR_SITE_DRIVE_KEY, {}).get(drive_id)
+
+    def drive_item_operation(self, item):
+        if "deleted" in item:
```

Comment:
It seems we should check `@removed` for delete operation:
>Removed instances are represented by their id and an @removed object. The @removed object may include additional information about why the instance was removed. For example, "@removed": {"reason": "changed"}.

Ref: https://learn.microsoft.com/en-us/graph/delta-query-overview#resource-representation-in-the-delta-query-response

What about `OP_UPSERT`?

Reply:
I tried `OP_UPSERT` but it does not trigger ingest pipelines because ingest pipeline trigger only when you index or re-index documents. Do you a better way to handle this?

Reply:
> It seems we should check `@removed` for delete operation:
> 
> > Removed instances are represented by their id and an @removed object. The @removed object may include additional information about why the instance was removed. For example, "@removed": {"reason": "changed"}.
> 
> Ref: https://learn.microsoft.com/en-us/graph/delta-query-overview#resource-representation-in-the-delta-query-response

This is weird because this is what I'm getting when a doc gets deleted:

```
{
'@odata.type': '#microsoft.graph.driveItem', 
  'id': '01WKXJTCYBH6NXECRZ4ZFZVVQIJJ2LPEJD', 
  'size': 0, 
  'parentReference': {'driveType': 'documentLibrary', 'driveId': 'b!1SwS-wjro0ew8puOF4o0UwemLfYyGE9OgLUZ5HpVe-zwJGwDHhGKRaucdMWq3ZAv', 'id': '01WKXJTC56Y2GOVW7725BZO354PWSELRRZ', 'siteId': 'fb122cd5-eb08-47a3-b0f2-9b8e178a3453'}, 
  'deleted': {'state': 'deleted'}, 
  'file': {'hashes': {}}, '_id': '01WKXJTCYBH6NXECRZ4ZFZVVQIJJ2LPEJD', 'object_type': 'drive_item', '_timestamp': None
}
```

Reply:
> I tried `OP_UPSERT` but it does not trigger ingest pipelines because ingest pipeline trigger only when you index or re-index documents. Do you a better way to handle this?

TIL. I didn't find it documented in the [Bulk API doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html), but it does stated [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/ingest.html#add-pipeline-to-indexing-request):

>Use the `pipeline` query parameter to apply a pipeline to documents in [individual](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html) or [bulk](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html) indexing requests.
You can also use the `pipeline` parameter with the [update by query](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html) or [reindex](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html) APIs.